File: app/src/main/java/com/teixeira/subtitles/adapters/ProjectListAdapter.java
package com.teixeira.subtitles.adapters;

import android.animation.LayoutTransition;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ImageView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;
import com.blankj.utilcode.util.FileUtils;
import com.teixeira.subtitles.databinding.LayoutProjectItemBinding;
import com.teixeira.subtitles.models.Project;
import com.teixeira.subtitles.utils.UiUtils;
import com.teixeira.subtitles.utils.VideoUtils;
import java.util.Collections;
import java.util.List;

public class ProjectListAdapter extends RecyclerView.Adapter<ProjectListAdapter.VH> {

  private List<Project> projects = Collections.emptyList();

  private final ProjectListener listener;

  public ProjectListAdapter(ProjectListener listener) {
    this.listener = listener;
  }

  @Override
  public VH onCreateViewHolder(ViewGroup parent, int viewType) {
    return new VH(
        LayoutProjectItemBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false));
  }

  @Override
  public void onBindViewHolder(VH holder, int position) {
    LayoutProjectItemBinding binding = holder.binding;
    Project project = projects.get(position);

    binding.videoIcon.setImageBitmap(VideoUtils.getVideoThumbnail(project.getVideoPath()));
    binding.name.setText(project.getName());
    binding.videoName.setText(FileUtils.getFileName(project.getVideoPath()));

    binding.getRoot().setOnClickListener(v -> listener.onProjectClickListener(v, project));
    binding
        .getRoot()
        .setOnLongClickListener(
            v -> {
              toggleOptionsVisibility(binding.optionsContainer, binding.chevron);
              return true;
            });
    binding.chevron.setOnClickListener(
        v -> toggleOptionsVisibility(binding.optionsContainer, binding.chevron));

    binding.options.editOption.setOnClickListener(
        v -> listener.onProjectOptionClickListener(v, project));

    binding.options.deleteOption.setOnClickListener(
        v -> listener.onProjectOptionClickListener(v, project));

    LayoutTransition layoutTransition = new LayoutTransition();
    layoutTransition.enableTransitionType(LayoutTransition.CHANGING);
    binding.getRoot().setLayoutTransition(layoutTransition);
  }

  @Override
  public int getItemCount() {
    return projects.size();
  }

  public void setProjects(@NonNull List<Project> projects) {
    DiffUtil.DiffResult result =
        DiffUtil.calculateDiff(new ProjectDiffCallback(this.projects, projects));
    this.projects = projects;
    result.dispatchUpdatesTo(this);
  }

  private void toggleOptionsVisibility(FrameLayout optionsContainer, ImageView chevron) {
    optionsContainer.setVisibility(
        optionsContainer.getVisibility() == View.VISIBLE ? View.GONE : View.VISIBLE);
    UiUtils.rotateView(chevron, optionsContainer.getVisibility() == View.VISIBLE ? 0.0f : -90.0f);
  }

  public interface ProjectListener {

    void onProjectClickListener(View view, Project project);

    void onProjectOptionClickListener(View view, Project project);
  }

  class VH extends RecyclerView.ViewHolder {
    private final LayoutProjectItemBinding binding;

    public VH(LayoutProjectItemBinding binding) {
      super(binding.getRoot());
      this.binding = binding;
    }
  }

  class ProjectDiffCallback extends DiffUtil.Callback {

    private final List<Project> oldList, newList;

    private ProjectDiffCallback(List<Project> oldList, List<Project> newList) {
      this.oldList = oldList;
      this.newList = newList;
    }

    @Override
    public int getOldListSize() {
      return oldList.size();
    }

    @Override
    public int getNewListSize() {
      return newList.size();
    }

    @Override
    public boolean areItemsTheSame(int oldPosition, int newPosition) {
      return areContentsTheSame(oldPosition, newPosition);
    }

    @Override
    public boolean areContentsTheSame(int oldPosition, int newPosition) {
      Project oldProject = oldList.get(oldPosition);
      Project newProject = newList.get(newPosition);

      return oldProject.getProjectPath().equals(newProject.getProjectPath())
          && oldProject.getVideoPath().equals(newProject.getVideoPath())
          && oldProject.getName().equals(newProject.getName());
    }
  }
}

File: app/src/main/java/com/teixeira/subtitles/adapters/ProjectListAdapter.kt
/*
 * This file is part of SubTypo.
 *
 * SubTypo is free software: you can redistribute it and/or modify it under the terms of
 * the GNU General Public License as published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * SubTypo is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with SubTypo.
 * If not, see <https://www.gnu.org/licenses/>.
 */

package com.teixeira.subtitles.adapters

import android.animation.LayoutTransition
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.ImageView
import androidx.core.view.isVisible
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.teixeira.subtitles.databinding.LayoutProjectItemBinding
import com.teixeira.subtitles.models.Project
import com.teixeira.subtitles.utils.UiUtils
import com.teixeira.subtitles.utils.VideoUtils

class ProjectListAdapter(
  val onProjectClick: (view: View, project: Project) -> Unit,
  val onProjectOptionClick: (view: View, project: Project) -> Unit,
) : RecyclerView.Adapter<ProjectListAdapter.ProjectViewHolder>() {

  private var projects = listOf<Project>()

  inner class ProjectViewHolder(val binding: LayoutProjectItemBinding) :
    RecyclerView.ViewHolder(binding.root)

  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ProjectViewHolder {
    return ProjectViewHolder(
      LayoutProjectItemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
    )
  }

  override fun onBindViewHolder(holder: ProjectViewHolder, position: Int) {
    holder.binding.apply {
      val project = projects[position]

      videoIcon.setImageBitmap(VideoUtils.getVideoThumbnail(project.videoPath))
      videoName.setText(project.videoName)
      name.setText(project.name)

      root.setLayoutTransition(
        LayoutTransition().apply { enableTransitionType(LayoutTransition.CHANGING) }
      )

      root.setOnClickListener { onProjectClick(it, project) }
      root.setOnLongClickListener {
        toggleOptionsVisibility(optionsContainer, chevron)
        true
      }
      chevron.setOnClickListener { toggleOptionsVisibility(optionsContainer, chevron) }

      options.editOption.setOnClickListener { onProjectOptionClick(it, project) }

      options.deleteOption.setOnClickListener { onProjectOptionClick(it, project) }
    }
  }

  override fun getItemCount(): Int = projects.size

  fun submitList(newProjects: List<Project>) {
    val result = DiffUtil.calculateDiff(ProjectDiffCallback(projects, newProjects))
    this.projects = newProjects
    result.dispatchUpdatesTo(this)
  }

  private fun toggleOptionsVisibility(optionsContainer: FrameLayout, chevron: ImageView) {
    optionsContainer.isVisible = !optionsContainer.isVisible
    UiUtils.rotateView(chevron, if (optionsContainer.isVisible) 0.0f else -90.0f)
  }

  class ProjectDiffCallback(val oldList: List<Project>, val newList: List<Project>) :
    DiffUtil.Callback() {

    override fun getOldListSize(): Int = oldList.size

    override fun getNewListSize(): Int = newList.size

    override fun areItemsTheSame(oldPosition: Int, newPosition: Int): Boolean {
      return oldList[oldPosition] == newList[newPosition]
    }

    override fun areContentsTheSame(oldPosition: Int, newPosition: Int): Boolean {
      val oldProject = oldList[oldPosition]
      val newProject = newList[newPosition]

      return oldProject.id == newProject.id &&
        oldProject.name == newProject.name &&
        oldProject.videoPath == newProject.videoPath
    }
  }
}
